crates/wasmtime/src/sig_registry.rs

```diff
@@ -36,15 +49,18 @@ impl SignatureRegistry {
             hash_map::Entry::Vacant(entry) => {
                 // Keep `signature_hash` len under 2**32 -- VMSharedSignatureIndex::new(std::u32::MAX)
                 // is reserved for VMSharedSignatureIndex::default().
-                assert_lt!(
-                    len,
-                    std::u32::MAX as usize,
+                assert!(
+                    len < std::u32::MAX as usize,
                     "Invariant check: signature_hash.len() < std::u32::MAX"
                 );
+                debug_assert_eq!(len, self.index_map.len());
                 let index = VMSharedSignatureIndex::new(u32::try_from(len).unwrap());
+                self.index_map.push(Entry {
+                    wasm: wasm.clone(),
+                    native: native.clone(),
```

Comment:
Since this function has to take ownership, we might as well make the caller do the clone (if necessary) so that we can avoid clones when they aren't needed. That is, revert the change from taking `&WasmFuncType` and `&ir::Signature` back to taking `WasmFuncType` and `ir::Signature`.

Reply:
This is sort of the `Entry` problem where it's problematic that getting an `entry` requires `wasm.clone()` (should probably just so a `get` followed by an `insert`, but also that the `native` isn't always inserted here (although it sometimes is)

If we think this is an issue I could change the arguments to `impl Into<Cow<'_, ...>>`, but I was hoping to optimize more for the cache hit use case here rather than the "need to insert case". Do you think I should try to go the `Cow` route though?

Reply:
If it doesn't require bending over backwards too much, it would be nice.

Reply:
Hm it turns out that `Into<Cow>` doesn't have as many implementations as I thought it did so I don't think this will work easily. Digging in more though I'm realizing that we're doing a number of unnecessary allocations in a number of places. In general a `Module` has no need to store the native signature types (it only needs the wasm signatures).

Would you be ok landing this as-is and I can have a follow-up which I think makes all the callers only have `&WasmFuncType` anyway so there's no gain to be had by passing in an owned value?

Reply:
Yeah, for sure